### ISSUE TITLE
Transitioning to Julian datasets and data loaders

### ISSUE THREAD

**Issue by darsnack** [FluxML/ML-Coordination-Tracker#7]
I wanted to add to this the general discussion from Zulip (linked below). It's best we transfer concrete suggestions to here so it's available for everyone to contribute. The plan has evolved into two "milestones" — short-term and long-term. I'll summarize both below.

Short-term:
- [x] Implement iterable and map-like datasets à la PyTorch
- [x] Implement a DataLoader interface mimicking PyTorch
- The goal here is to take PyTorch implementations as-is to get something up and running, since this is a desperately needed functionality in the ecosystem
- Being tracked here: #5 

Long-term:
- [x] Transfer iterable dataset interface to [Base iterator interface](https://docs.julialang.org/en/v1/manual/interfaces/#man-interface-iteration-1) [1]
- [x] Transfer the map-like dataset interface to [Base indexable collections](https://docs.julialang.org/en/v1/base/collections/#Indexable-Collections-1) [1]
- The DataLoader interface doesn't necessarily need to change, but long term we should make sure that concrete implementations of the interface take advantage of things like Base.Random and Distributions for sampling

Relevant Zulip topics:
- https://julialang.zulipchat.com/#narrow/stream/237432-ml-ecosystem-coordination/topic/Datasets
- https://julialang.zulipchat.com/#narrow/stream/237432-ml-ecosystem-coordination/topic/DataLoaders

----
[1]: the Base interfaces may not be perfect for our needs, so we might need to build off of them

**Comment by darsnack** [comment]
cc @opus111 @SomTambe

**Comment by lorenzoh** [comment]
I think the LearnBase.jl is being established pretty nicely, with [DataLoaders.jl](https://github.com/lorenzoh/DataLoaders.jl) and [FastAI.jl](https://github.com/FluxML/FastAI.jl) fully embracing it.

As an update:
- FastAI.jl also includes *all* datasets from the [fastai dataset repository](https://course.fast.ai/datasets) and prototype data containers for files and tables which will be moved to MLDatasets.jl when stable (see [this PR](https://github.com/JuliaML/MLDatasets.jl/pull/56).
- Additionally, there are useful [data container transformations](https://github.com/FluxML/FastAI.jl/blob/master/src/datasets/transformations.jl) that, once stable, will be merged into [MLDataPattern.jl](https://github.com/JuliaML/MLDataPattern.jl).

**Comment by darsnack** [comment]
I think most of the original issue has effectively been completed by adopting JuliaML/LearnBase.jl. In addition to the dataset transfer and data container transformations above, there are some short-term objectives that need to be met:
- Updating/refreshing LearnBase.jl's interface so that future development is easier: https://github.com/JuliaML/LearnBase.jl/pull/44
    - Updating MLDataPattern to reflect these changes: https://github.com/JuliaML/MLDataPattern.jl/pull/50
    - Updating MLLabelUtils to reflect these changes: https://github.com/JuliaML/MLLabelUtils.jl/pull/44

**Comment by lorenzoh** [comment]
Yeah, I didn't want to close this, just to give an update on what's been happening 👍

**Comment by darsnack** [comment]
Closing this since we either have completed the sub-issue as part of the LearnBase.jl -> MLUtils.jl transition, or we have a tracking issue on MLUtils.jl / MLDatasets.jl itself.
